Added an option for PPO to train on incomplete episodes.

PeterZhizhin opened the pull request:
As discussed in #30, right now PPO discards all rewards that were
found on not completed episodes. This commit addresses this issue and
adds an option to train on episodes that were not completed.

Also added tests to ensure that loss is not zero when training on
incomplete episodes depending on the flags.

danijar commented:
Yes, probably. I think TensorFlow doesn't have a filter() which I needed to get only the tensor-types elements of a distribution object's `dist.parameters`.

PeterZhizhin committed:
Added an option for PPO to train on incomplete episodes.

As discussed in #30, right now PPO discards all rewards that were
found on not completed episodes. This commit addresses this issue and
adds an option to train on episodes that were not completed.

Also added tests to ensure that loss is not zero when training on
incomplete episodes depending on the flags.

ebrevdo commented on tf_agents/agents/ppo/ppo_agent.py:
```diff
@@ -118,6 +118,7 @@ def __init__(self,
                normalize_rewards=True,
                reward_norm_clipping=10.0,
                normalize_observations=True,
+               train_only_on_full_episodes=True,
```
@oars should we make the default false here?

PeterZhizhin replied:
I would also like to know whether there is a reason behind this `valid_mask`?
I didn't see this implemented in any common RL libraries such as OpenAI Baselines or Ray RLlib.

oars replied:
The mask is needed because of the way data is added to the RB. In general observers are not filtering transitions between `step_type.LAST` and `step_type.FIRST`. I also don't think all agents are handling this correctly, in general we need to fix this.

@ebrevdo defaulting to False would definetly be the safest option to make sure behavior doesn't change for users.

@sguada Know if any other parts of PPO rely on the experience being full episodes?

sguada replied:
I think what we need is to bootstrap the end of episodes using the value function if it is not an end of episode.

But we still need to use the mask to avoid invalid transitions between `LAST` and `FIRST`.

Rename to `train_only_on_partial_episodes=True,`

PeterZhizhin replied:
New commits solve the described issues.

ebrevdo commented on tf_agents/agents/ppo/ppo_agent.py:
```diff
@@ -487,7 +492,10 @@ def _train(self, experience, weights):
+    if self._train_only_on_full_episodes:
+      valid_mask = ppo_utils.make_timestep_mask(next_time_steps)
+    else:
+      valid_mask = tf.ones(next_time_steps.step_type.shape)
```
is there a way to make valid_mask None here and respect that throughout the rest of the code, thereby avoiding an extra useless tensor?

PeterZhizhin committed:
Allow weights to be None for losses in PPOAgent.

PeterZhizhin commented:
I have resolved the issue with a potentially unused tensor. Waiting for @oars to answer.

sguada commented on tf_agents/agents/ppo/ppo_agent.py:
```diff
@@ -487,11 +492,13 @@ def _train(self, experience, weights):
         experience.observation, experience.step_type, policy_state=policy_state)
     value_preds = tf.stop_gradient(value_preds)
 
-    valid_mask = ppo_utils.make_timestep_mask(next_time_steps)
```
pass `train_only_on_partial_episodes` to `make_timestep_mask` and build an appropriate mask.

keep `valid_mask` in the rest of code.

PeterZhizhin replied:
Used `train_on_partial_episodes` instead of `train_only_on_partial_episodes` as it should also work in episodic envs as well.

Bootstrapped the reward using the value function as well.

sguada reviewed (commented):
```
def make_timestep_mask(batched_next_time_step, partial_episodes=False):
  """Create a mask for final incomplete episodes and episode transitions.
  Args:
    batched_next_time_step: Next timestep, doubly-batched
      [batch_dim, time_dim, ...].
    partial_episodes: Bool, when True allow partial episodes.
  Returns:
    A mask, type tf.float32, that is 0.0 for all between-episode timesteps
      (batched_next_time_step is FIRST), or where the episode is
      not compelete, so the return computation would not be correct.
  """
  # 1.0 for timesteps of all complete episodes. 0.0 for incomplete episode at
  #   the end of the sequence.
  if partial_episodes:
   episode_is_complete = tf.ones_like(batched_next_time_step.is_last(), dtype=tf.bool)
  else:
   episode_is_complete = tf.cumsum(
      tf.cast(batched_next_time_step.is_last(), tf.float32),
      axis=1,
      reverse=True) > 0

  # 1.0 for all valid timesteps. 0.0 where between episodes.
  not_between_episodes = ~batched_next_time_step.is_first()

  return tf.cast(episode_is_complete & not_between_episodes, tf.float32)
```

PeterZhizhin committed:
Merge branch 'master' into train_in_non_episodic

PeterZhizhin committed:
Training on partial episodes now bootstraps reward

As discussed in the rewiew, we should boostrap the reward function using
our value function predictions for training on partial episodes.

PeterZhizhin committed:
Added comment on bootstrapping value function & the last episode really being the last episode

PeterZhizhin committed:
Removed "weight is None" checks in loss functions.

As we discussed that we are always using the weight mask, then there is
no need in these checks.

PeterZhizhin committed:
Merge branch 'master' of https://github.com/tensorflow/agents into train_in_non_episodic

PeterZhizhin commented:
I see that there is an error while migrating the change. Could you please clarify what is the problem?

sguada commented:
You probably need to sync your branch to the latest code and merge any conflicts.

andreas-repp commented:
@PeterZhizhin: Can you update your branch so they can merge it? I'd really like to see this implemented. Thanks already for providing this PR in the first place!

tfboyd commented:
@sguada @ebrevdo   The conflicts are not that hard to resolve.  Assuming the person does not resond (it has been a long time); I am happy to fix the conflicts and finish this off if this is something we want.  It looks like it was almost done.

ebrevdo commented on tf_agents/agents/ppo/ppo_agent.py:
```diff
@@ -164,6 +165,9 @@ def __init__(self,
       reward_norm_clipping: Value above an below to clip normalized reward.
       normalize_observations: If true, keeps moving mean and variance of
         observations and normalizes incoming observations.
+      train_on_partial_episodes: If True (default False), then data from
```
i wonder if we can just always train on partial episodes.  @oars wdyt?  can we get rid of this arg and always assume true?

oars replied:
I think it should be fine. Previous use cases will still work as they are feeding full episodes already.

ebrevdo commented:
This looks great; I just want to see if we can always train from partial episodes.  If so we can get rid of one of the init args.

tfboyd commented:
I am 95% sure the copybara process is failing because of the branch conflict.  The PR branch does not have /docs, which is expected by copybara, which would be fine but I suspect because it is not merging due to the conflict the branch copybara is seeing does not have /docs.

tfboyd commented:
I am making a new PR to fix the conflict and preserve the commits and credit for the original author.  Should be done in a few minutes....fingers crossed.

tfboyd commented:
Already I merged Peter's commits and made a new PR fixing the conflicts.  I am running the unit tests as well.  Please go here.  https://github.com/tensorflow/agents/pull/306
